sm_ipt/sm_ipt.c

```diff
@@ -0,0 +1,450 @@
+#define WORD_BITS (8 * sizeof(uint32_t))
+
+#if defined(CONFIG_SM_IPT_NUM_BLOCKS_32)
+typedef uint32_t mask_t;
+typedef int32_t smask_t;
+#define NUM_BLOCKS 32
+#define mask_clz sm_ipt_os_clz32
+#elif defined(CONFIG_SM_IPT_NUM_BLOCKS_64)
+typedef uint64_t mask_t;
+typedef int64_t smask_t;
+#define NUM_BLOCKS 64
+#define mask_clz sm_ipt_os_clz64
+#else
+#error Number of shared dynamic memory blocks is not configured.
+#endif
+
+#define ALLOCABLE_MULTIPLY (WORD_SIZE * NUM_BLOCKS)
+#define QUEUE_INDEX_MASK (NUM_BLOCKS - 1)
+
+/* Worst case scenario: NUM_BLOCKS count of messages waiting to be sent and NUM_BLOCKS count of
+ * buffers waiting to be freed + 1 item used as indicator of full queue
+ */
+#define QUEUE_ITEMS_CNT (2 * NUM_BLOCKS + 1)
+
+#define memory_corrupted_error() sm_ipt_os_fatal(); \
+	SM_IPT_ERR("Shared memory corrupted"); \
+	SM_IPT_ASSERT(0)
+
+
+struct sm_ipt_shm_queue {
+	uint32_t queue_tx;
+	uint32_t queue_rx;
+	uint8_t queue[QUEUE_ITEMS_CNT];
+	uint8_t handshake;
+	uint16_t _padding; /* Care should be taken to align allocable to word boundary */
+	uint8_t allocable[];
+};
+
+
+static void free_mask_set(struct sm_ipt_ctx *ctx, mask_t mask)
+{
+	sm_ipt_os_atomic_or(&ctx->free_mask[0], mask);
+
+	if (NUM_BLOCKS > 32) {
+		sm_ipt_os_atomic_or(&ctx->free_mask[1], (uint64_t)mask >> 32);
+	}
+
+	SM_IPT_OS_MEMORY_BARRIER();
+}
+
+static bool free_mask_unset(struct sm_ipt_ctx *ctx, mask_t mask)
+{
+	uint32_t old;
+	uint32_t mask_part;
+
+	mask_part = (uint32_t)mask;
+	old = sm_ipt_os_atomic_and(&ctx->free_mask[0], ~mask_part);
+	if ((old & mask_part) != mask_part) {
+		sm_ipt_os_atomic_or(&ctx->free_mask[0], old & mask_part);
+
+		return false;
+	}
+
+	if (NUM_BLOCKS > 32) {
+		mask_part = (uint32_t)((uint64_t)mask >> 32);
+		old = sm_ipt_os_atomic_and(&ctx->free_mask[1], ~mask_part);
+
+		if ((old & mask_part) != mask_part) {
+			sm_ipt_os_atomic_or(&ctx->free_mask[1], old & mask_part);
+			sm_ipt_os_atomic_or(&ctx->free_mask[0], (uint32_t)mask);
+			return false;
+		}
+	}
+
+	SM_IPT_OS_MEMORY_BARRIER();
+
+	return true;
+}
+
+static mask_t free_mask_get(struct sm_ipt_ctx *ctx)
+{
+	mask_t mask;
+
+	mask = sm_ipt_os_atomic_get(&ctx->free_mask[0]);
+
+	if (NUM_BLOCKS > 32) {
+		mask = (uint64_t)sm_ipt_os_atomic_get(&ctx->free_mask[1]) << 32;
+	}
+
+	return mask;
+}
+
+static void free_mask_init(struct sm_ipt_ctx *ctx)
+{
+	free_mask_set(ctx, ~(mask_t)0);
+}
+
+static mask_t calc_mask(size_t blocks, size_t index)
+{
+	mask_t mask;
+
+	mask = (mask_t)1 << (NUM_BLOCKS - 1);        /* 100000000... */
+	mask = (smask_t)mask >> (blocks - 1);        /* 111000000... (e.g. blocks = 3) */
+	mask = mask >> index;                        /* 000011100... (e.g. index = 4) */
+
+	return mask;
+}
+
+void sm_ipt_alloc_tx_buf(struct sm_ipt_ctx *ctx, uint8_t **buf, size_t len)
+{
+	/* Actual allocated memory: | 32-bit size | data | padding | */
+	size_t blocks = (len + (WORD_SIZE + ctx->out.block_size - 1)) / ctx->out.block_size;
```

Comment:
You often use `WORD_SIZE`. In this case it looks like you reserve a word for some special data. Maybe you could use a define for it. I assume that you reserve 4 byte here for data length so maybe:
`#define DATA_LENGTH_FIELD_SIZE WORD_SIZE`

or something similar